lib/hl7/parser.ex

```diff
@@ -0,0 +1,118 @@
+defmodule HL7.Parser do
+  @carriage_return 4
+  @field 3
+  @repetition 2
+  @component 1
+  @sub_component 0
+
+  def parse(<<"MSH|^~\\&", rest::binary()>> = text) do
+    tokenize(rest, text, 8, 0, ["^~\\&", @field, "|", @field, "MSH"])
+    |> to_lists()
+  end
+
+  def tokenize(<<char::binary-size(1), rest::binary>>, original, skip, len, acc) do
```

Comment:
Do all of these functions need to be public?

Reply:
Curious why we preserve `original` and pass it around with a `skip` instead of chopping it down as we progress? (Which would obviously lead to a name change.)

Reply:
The `original` is used to grab strings out (that preceded where we are in the parse). This is code I blatantly ripped off from the Jason implementation ;)

Reply:
I looked for something like that, but I'm still missing it. Where does that happen?

Reply:
```
 defp tokenize_terminator(text, original, skip, len, acc, terminator) do
    string = binary_part(original, skip, len)
    tokenize(text, original, skip + len + 1, 0, [terminator, string | acc])
  end
  ```
The 'binary_part` call pulls it out (we keep incrementing skip and len as we go through the text).

Reply:
I don't see `original` referenced anywhere without `binary_part` and `skip`. Why not chop its head off instead of incrementing `skip` and passing it around as a marker of where to start?

Reply:
rewritten to be much faster (new grossness!)

Reply:
ps - the way it uses the data here to cut strings is identical to (ie stolen from) Jason. It's apparently a perf thing the way data is stored and handled.